### ISSUE TITLE
User exceptions no longer propagate out of Python operator wrappers

### ISSUE THREAD

**Issue by e4lam** [swig/swig#1783]
I've run into an issue with https://github.com/swig/swig/commit/687cf9c9c1b886ef4a828b2c29610183e9fef26d (by @wsfulton ). In our C++ `operator==()` methods, we might throw exceptions that used to be raised back in Python land, but this no longer works in SWIG 4 due to this change. We've just implemented exceptions as suggested in the docs as:
```
%exception {
    try
    {
        $action
    }
    catch(...)
    {
        hom::raise_swig_exception(); // custom exception translation
        SWIG_fail;
    }
}
```
The problem is that after https://github.com/swig/swig/commit/687cf9c9c1b886ef4a828b2c29610183e9fef26d, the generated wrapper code now looks something like this:
```
SWIGINTERN PyObject *_wrap_SwigPyIterator_equal(PyObject *SWIGUNUSEDPARM(self), PyObject *args) {
    ....
  {
    try
    {
      result = (bool)((swig::SwigPyIterator const *)arg1)->operator ==((swig::SwigPyIterator const &)*arg2);
    }
    catch(...)
    {
      hom::raise_swig_exception(); // custom exception translation
      SWIG_fail;
    }
  }
  resultobj = SWIG_From_bool(static_cast< bool >(result));
  return resultobj;
fail:
  PyErr_Clear();
  Py_INCREF(Py_NotImplemented);
  return Py_NotImplemented;
}
```
So basically, any exception that the C++ code throws, it just ends up in the `fail:` case and then we end up eating the exception by returning `Py_NotImplemented` instead of returning `NULL` which would let Python know that an exception happened.

**Comment by e4lam** [comment]
I did a quick fix by

**Comment by vadz** [comment]
This got me confused for a moment, but this doesn't really seem to have anything to do with GitHub issue \#303, just with the commit 687cf9c which got (wrongly) referenced there. But FWIW the new behaviour does seem wrong to me, and it looks like we ought just return `NotImplemented` when there is no such overload for `==` at all, rather than when it throws.

I wonder if it's possible to reset `python:maybecall` on the affected operators?

**Comment by e4lam** [comment]
I did a quick fix by:
```
@@ -303,7 +303,7 @@
     catch(...)
     {
         hom::raise_swig_exception();
-        SWIG_fail;
+        return nullptr;
     }
 }
```
But I don't know if that's safe?

**Comment by e4lam** [comment]
@vadz Sorry! That's my bad. I'll fix the description!

**Comment by e4lam** [comment]
@vadz Yeah, it kinda feels like the wrong fix. There's really only one failure condition in the generated wrapper code right now so I can see why it was done there. I don't know if it's easy to change the other failure conditions to do this instead.

I'm not sure what you mean by:
> I wonder if it's possible to reset python:maybecall on the affected operators?

Is this something I can inside the `%exception` ?

**Comment by vadz** [comment]
> I'm not sure what you mean by:
> 
> > I wonder if it's possible to reset python:maybecall on the affected operators?
> 
> Is this something I can inside the `%exception` ?

No, this would need to be done (if possible) in your `.i` file. Sorry, this could be totally wrong, but from looking at that commit, it seems that this behaviour is only activated for the methods with `python:maybecall` feature active. And `pyopers.swg` activates it for all operators. So I thought that perhaps you could do

    %feature("python:maybecall", "0") *::operator==;

to turn it off.

**Comment by e4lam** [comment]
@vadz Ah right, but then this means we also undo the whole intended fix of avoiding the TypeError exception.

**Comment by vadz** [comment]
Yes, this is just a (possible) workaround, not the real fix.

**Comment by e4lam** [comment]
@vadz What about doing `return nullptr;` instead of `SWIG_fail;` inside the `%exception` ?

**Comment by vadz** [comment]
I suspect this is not going to work, at least if the same `%exception` code is used with all functions and not just the operators, but I'm not really sure.

**Comment by e4lam** [comment]
Yeah, I took a close look and found some cases where doing this will definitely leak memory.

**Comment by ojwb** [comment]
The docs from the commit linked above say:

>Operator overloading is implemented in the <tt>pyopers.swg</tt> library file.
In particular overloaded operators are marked with the <tt>python:maybecall</tt> feature, also known as <tt>%pythonmaybecall</tt>.
This feature forces SWIG to generate code that return an instance of Python's <tt>NotImplemented</tt>
instead of raising an exception when the comparison fails, that is, on any kind of error.
This follows the guidelines in <a href="https://www.python.org/dev/peps/pep-0207/">PEP 207 - Rich Comparisons</a> and <a href="https://docs.python.org/3/library/constants.html#NotImplemented">NotImplemented Python constant</a>.

However the _"on any kind of error"_ part doesn't seem to be supported by either of the linked Python docs - the second one seems clearest and says:

> A special value which should be returned by the binary special methods (e.g. [__eq__()](https://docs.python.org/3/reference/datamodel.html#object.__eq__), [__lt__()](https://docs.python.org/3/reference/datamodel.html#object.__lt__), [__add__()](https://docs.python.org/3/reference/datamodel.html#object.__add__), [__rsub__()](https://docs.python.org/3/reference/datamodel.html#object.__rsub__), etc.) to indicate that the operation is not implemented with respect to the other type; may be returned by the in-place binary special methods (e.g. [__imul__()](https://docs.python.org/3/reference/datamodel.html#object.__imul__), [__iand__()](https://docs.python.org/3/reference/datamodel.html#object.__iand__), etc.) for the same purpose.

My reading is this should only be done when comparing the types isn't implemented, rather than _"on any kind of error"_.

@wsfulton This was your change and seems to be a regression for people (like the reporter) who were throwing other exceptions from comparison operators.

**Comment by wsfulton** [comment]
You can specify a different %exception for different methods, so use the workaround for just the operators you need, such as:
```swig
%exception *::operator== {
...
    catch(...)
    {
        return nullptr;
...
```
BTW, SwigPyIterator::equal, which correlates to the generated code at the beginning, does not have %pythonmaybecall set on it by default, you must have set that yourself.

**Comment by wsfulton** [comment]
The generated code is now:
```c++
...
  resultobj = SWIG_From_bool(static_cast< bool >(result));
  return resultobj;
fail:
  if (PyErr_Occurred() && !PyErr_ExceptionMatches(PyExc_TypeError)) {
    return NULL;
  }
  PyErr_Clear();
  Py_INCREF(Py_NotImplemented);
  return Py_NotImplemented;
}
```
so that it only converts `TypeError` into `NotImplemented` as this is the exception SWIG throws when the types are not appropriate for the wrapper.

**Comment by e4lam** [comment]
Thanks @wsfulton ! And I forgot to thank @vadz (thanks!) for the workaround which we had used to workaround this issue as well.
